Add generic gpio leds driver with sample

KAGA164 opened the pull request:
This PR adds leds gpio driver which is used to control the LED devices attached to the GPIO pins through the LED API. The driver support `led_on` , `led_off` and `led_blink`. The blink feature is realized by using software timer and can be disabled from compilation. Driver can handle all LEDs defined in DTS which contains id field. The `led_set_brightness` is not supported and I see no need to add it in case of the GPIO.

There are plans in feature to extend this driver to support power management.

Generation of the common static initializers for child-bindings is needed for this driver. It is provided in the #21393 PR. The first three commits come from there and driver will be rebased if #21393 will be merged, so them are not a part of this PR. Also driver needs additional id field in the DTS led description. This filed is not required to to ensure backward compatibility and it should be added to use this driver.

Commit:
board: LED id for nrf52840_pca10056

Add LED ID for nrf52840_pca10056 board.
This id is required for gpio leds driver.

Signed-off-by: Kamil Gawor <Kamil.Gawor@nordicsemi.no>

Commit:
samples: driver: Add LED GPIO sample

This commit provides a new sample which demonstrate
how to use the GPIO LED drivers.

Signed-off-by: Kamil Gawor <Kamil.Gawor@nordicsemi.no>

KAGA164 commented:
@anangl Please look at this

Mani-Sadhasivam commented on drivers/led/CMakeLists.txt:
```diff
@@ -4,5 +4,6 @@ zephyr_sources_ifdef(CONFIG_HT16K33 ht16k33.c)
 zephyr_sources_ifdef(CONFIG_LP3943 lp3943.c)
 zephyr_sources_ifdef(CONFIG_LP5562 lp5562.c)
 zephyr_sources_ifdef(CONFIG_PCA9633 pca9633.c)
+zephyr_sources_ifdef(CONFIG_LED_GPIO gpio.c)
```
I'd name this file as `led_gpio` to match the config.

KAGA164 replied:
Ok sounds good. Changed

Mani-Sadhasivam commented on drivers/led/Kconfig.gpio:
```diff
@@ -0,0 +1,24 @@
+	 Support for led software blink based on
+	 system timers
+
+module = LEDS_GPIO
```
For consistency, let's call LED_GPIO everywhere.

KAGA164 replied:
Ok

Mani-Sadhasivam commented on drivers/led/Kconfig:
```diff
@@ -24,5 +24,6 @@ source "drivers/led/Kconfig.ht16k33"
 source "drivers/led/Kconfig.lp3943"
 source "drivers/led/Kconfig.lp5562"
 source "drivers/led/Kconfig.pca9633"
+source "drivers/led/Kconfig.gpio"
```
`drivers/led/Kconfig.led_gpio`

KAGA164 replied:
Changed

Mani-Sadhasivam commented on drivers/led/gpio.c:
```diff
@@ -0,0 +1,251 @@
+#include <sys/util.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(led_gpio, CONFIG_LEDS_GPIO_LOG_LEVEL);
```
Logging is broken as I'm getting lot of warnings.

KAGA164 replied:
I fix some issue with logging. I tested with debug log and without. You have to build sample for `nrf52840_pca10056` board. The DTS descriptions is only aligned for it in this PR.

Mani-Sadhasivam commented on drivers/led/gpio.c:
```diff
@@ -0,0 +1,251 @@
+#include <kernel.h>
+#include <led.h>
+#include <gpio.h>
+#include <device.h>
```
Sort includes accordingly

KAGA164 replied:
I put this in alphabetical order. Is this what you mean ?

Mani-Sadhasivam replied:
yeah

Mani-Sadhasivam commented on drivers/led/gpio.c:
```diff
@@ -0,0 +1,251 @@
+	return NULL;
+}
+
+static int gpio_led_set_no_delay(struct led_gpio_data *data, bool on)
```
Better to use `state` instead of `on` everywhere.

KAGA164 replied:
Ok changed

Mani-Sadhasivam commented on drivers/led/gpio.c:
```diff
@@ -0,0 +1,251 @@
+	data->blink_delay_off = delay_off;
+
+	LOG_DBG("%s blink start, delay on: %d, delay off: %d",
+		delay_on, delay_off);
```
Missing `data->label`

KAGA164 replied:
Right. Fixed

Mani-Sadhasivam commented on dts/bindings/gpio/gpio-leds.yaml:
```diff
@@ -18,3 +26,7 @@ child-binding:
+       id:
+          required: false
+          type: int
+          description: Id needed for GPIO LED driver to identify LED
```
`ID needed for GPIO LED driver to identify an LED`

KAGA164 replied:
Changed to` ID needed for the GPIO LED driver to identify an LED`

Mani-Sadhasivam commented on dts/bindings/gpio/gpio-leds.yaml:
```diff
@@ -6,6 +6,14 @@ title: GPIO LEDs
 description: GPIO LEDs parent node
 
 compatible: "gpio-leds"
+generate-child-initializer: true
```
Since we now have a dedicated LED GPIO driver, we should think about moving this under `led` and also fixing the basic sample applications to use this new driver. This can be tackled later on also.

KAGA164 replied:
Good point. I think this can be done after #21393 merge. Then I will know the final form of the DTS bindigs and geration. So I will do it while rebasing.

Mani-Sadhasivam commented on drivers/led/gpio.c:
```diff
@@ -0,0 +1,251 @@
+{
+	struct led_gpio_data *data;
+	u32_t delay;
+	bool led_on;
```
state?

KAGA164 replied:
Can be. Changed

Mani-Sadhasivam commented on samples/drivers/led_gpio/src/main.c:
```diff
@@ -0,0 +1,86 @@
+#include <zephyr.h>
+
+#define LED_DEV_NAME DT_GPIO_LEDS_LEDS_DRIVER
+#define NUM_LEDS 4
```
Why 4? Not all boards will have 4 LEDs. I'll suggest using 2.

KAGA164 replied:
Right. I reduce LEDs count to 2.

Mani-Sadhasivam commented on samples/drivers/led_gpio/src/main.c:
```diff
@@ -0,0 +1,86 @@
+
+	led_dev = device_get_binding(LED_DEV_NAME);
+	if (led_dev) {
+		printk("Found LED device %s\n", LED_DEV_NAME);
```
Please use logging subsystem.

KAGA164 replied:
Mostly the applications use the `printk` instead of logging subsystem.  I thought in app we should use the `printk`. I am not sure if it need to be changed

Mani-Sadhasivam commented on samples/drivers/led_gpio/src/main.c:
```diff
@@ -0,0 +1,86 @@
+		return;
+	}
+
+	LOG_INF("Testing leds");
```
Logging is not registered for this sample.

KAGA164 replied:
`printk` in sample ?

Mani-Sadhasivam replied:
I don't really know what the convention is (logging vs printk in samples). I'm fine with either one but was just making sure it is not broken and only one implementation is used.

Mani-Sadhasivam commented on samples/drivers/led_gpio/src/main.c:
```diff
@@ -0,0 +1,86 @@
+/*
```
As suggested above, it'd be better if we can fix the existing basic applications instead of creating a new one.

KAGA164 replied:
I thought about this but using the LEDs GPIO drivers needs DTS modification for every board. So this PR is only a sample how to align the board dts. I can align all Nordic boards after it will be merged. Every LED have to have id field in its descriptions. So if I modify this sample then it will be broken for not aligned board. That is why I decided to create the new one sample. By the way this simple basic blinky sample in not broken. I think using LED driver is not mandatory for all boards so only those who decided to align their board could use it.

Mani-Sadhasivam commented on samples/drivers/led_gpio/README.rst:
```diff
@@ -0,0 +1,31 @@
+
+The sample assumes that the LEDs are connected to GPIO lines.
+It is configured to work on boards that have defined the led0, led1, led2, led3
+aliases in their board devicetree description file.
```
Why aliases are required?

KAGA164 replied:
Aliases are probably more common to use in sample or application. I did change sample to use it.

Mani-Sadhasivam replied:
You might need to mention the use of `id` property also, since that is a requirement for the driver.

Mani-Sadhasivam commented on drivers/led/gpio.c:
```diff
@@ -0,0 +1,251 @@
+	const char *label;
+	u32_t id;
+	struct device *gpio_dev;
+#if defined(CONFIG_LED_SOFTWARE_BLINK)
```
Please split the software blinking functionality into a separate commit.

KAGA164 replied:
Ok. I splitted it but why You want me to do it ?

Mani-Sadhasivam replied:
Just to make it easy to bisect in future and moreover these 2 are different functionalities.

Mani-Sadhasivam reviewed (dismissed):
@KAGA164 Thanks for the work! I don't know if you have noticed or not, I tried to add this driver a while ago (#9511) but didn't get time to complete due to some missing dependencies. But happy to see this finally getting some momentum. 

Overall, I like the idea and implementation but there are some comments I have added. Please address those.

dbkinder commented on samples/drivers/led_gpio/README.rst:
```diff
@@ -0,0 +1,31 @@
+********
+
+The sample assumes that the LEDs are connected to GPIO lines.
+It is configured to work on boards that have defined the led0, led1, led2, led3
```
change to ``led2, and led3 aliases``

KAGA164 replied:
Ok. I removed led2 and led3. Sample can handle two leds now to ensure compatibility with more boards.

dbkinder commented on drivers/led/Kconfig.gpio:
```diff
@@ -0,0 +1,24 @@
+config LED_SOFTWARE_BLINK
+	bool "Software blink"
+	help
+	 Support for led software blink based on
```
documentation and help text should capitalize ``LED``

KAGA164 replied:
Ok. I think that I changed it in all places.

dbkinder commented on samples/drivers/led_gpio/README.rst:
```diff
@@ -0,0 +1,31 @@
+   :goals: build flash
+   :compact:
+
+After flashing the image to the board, the user LEDs on the board should act according to the given pattern.
```
The reel_board used in your example doesn't have four LEDs on the board?

KAGA164 replied:
Right I checked it. So I put there `nrf52840_pca10056` board which have aligned dts to use the LED GPIO driver. In future we can use `reel_board`.

Commit:
driver: Add LED GPIO driver

Add driver for controlling the LED devices attached to GPIO pins
which are declared in devicetree under leds node and generated as static
initializer. The driver takes care of configuring the pins to output
state before using it.

Signed-off-by: Kamil Gawor <Kamil.Gawor@nordicsemi.no>

Commit:
driver: gpio_led: Software LED blink

Add software LED blink feature for each
defined LED. Blink funcionality uses software timer.

Signed-off-by: Kamil Gawor <Kamil.Gawor@nordicsemi.no>

KAGA164 commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,251 @@
+	LOG_DBG("%s blink start, delay on: %d, delay off: %d",
+		data->label, delay_on, delay_off);
+
+	k_timer_start(&data->timer, 0, 0);
```
Yes it is needed here. I had to remove it accidentally.

KAGA164 commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,251 @@
+{
+	int err;
+
+	if (state) {
```
I wrote a new version of this function

KAGA164 commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,251 @@
+}
+
+static struct led_gpio_data led_data[] = DT_INST_0_GPIO_LEDS_INIT;
+static struct led_gpio_cfg led_cfg = {
```
Ok. Added

KAGA164 commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,251 @@
+struct led_gpio_data {
+	struct gpio_cfg gpio;
+	const char *label;
+	u32_t id;
```
Moved to constans data structure. This data is in flash now

KAGA164 commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,251 @@
+#include <device.h>
+#include <gpio.h>
+#include <kernel.h>
+#include <led.h>
```
Ok.

anangl commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,261 @@
+
+	err = gpio_pin_write(data->gpio_dev, data->cfg->gpio.pin,
+		state ? ((data->cfg->gpio.polarity & GPIO_INT_ACTIVE_HIGH) != 0) :
+		((data->cfg->gpio.polarity & GPIO_INT_ACTIVE_HIGH) == 0));
```
What about using someting like this:
```
	u32_t pin_value;

	if (data->cfg->gpio.polarity & GPIO_INT_ACTIVE_HIGH) {
		pin_value = (state ? 1 : 0);
	} else {
		pin_value = (state ? 0 : 1);
	}
```
Could that make things a bit clearer (and shorten the lines so that checkpatch no longer complains)?

KAGA164 replied:
I changed it also in init function.

KAGA164 commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,261 @@
+static const struct led_dts_cfg led_dts[] = DT_INST_0_GPIO_LEDS_INIT;
+static struct led_gpio_data led_data[ARRAY_SIZE(led_dts)];
+static const struct led_gpio_cfg led_cfg = {
+	.led_cnt = ARRAY_SIZE(led_data),
```
Ok changed

anangl commented:
#21393 needs to go in first.

carlescufi commented:
@Mani-Sadhasivam please re-review
@KAGA164 please rebase and fix CI

Mani-Sadhasivam reviewed (approved):
LGTM.

b0661 commented on dts/bindings/gpio/gpio-leds.yaml:
```diff
@@ -15,3 +23,7 @@ child-binding:
           required: true
           type: string
           description: Human readable string describing the device (used by Zephyr for API name)
+       id:
```
Looking at https://www.kernel.org/doc/Documentation/devicetree/bindings/leds/leds-gpio.txt the property **id** could be named **function** instead. **functions** seems to have the same intention.

simonguinot commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,171 @@
+
+		/* Set all leds pin as output */
+		err = gpio_pin_configure(data[i].gpio_dev, cfg->led[i].gpio.pin,
+					 GPIO_DIR_OUT);
```
Hi @KAGA164 ,

I think you can pass the GPIO_INT_ACTIVE_{LOW,HIGH} flag to gpio_pin_configure(). This way you can call gpio_pin_set() directly with the logical value. This will handle the polarity directly. Maybe you also need to pass all the other GPIO flags found in the gpio DT property.

simonguinot commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,171 @@
+	.led = led_dts
+};
+
+DEVICE_AND_API_INIT(gpio_led, DT_INST_0_GPIO_LEDS_DRIVER,
```
Maybe it is possible to support all the gpio-led instances ? With this new gpio-led controller notion (with LEDs numbering), someone may want to group the GPIO LEDs into dedicated nodes.

simonguinot commented on drivers/led/led_gpio.c:
```diff
@@ -24,6 +27,13 @@ struct led_dts_cfg {
 struct led_gpio_data {
 	struct device *gpio_dev;
 	const struct led_dts_cfg *cfg;
+#if defined(CONFIG_LED_SOFTWARE_BLINK)
```
Maybe it would be interesting to move this software blink support directly into the LED API. This way every LED controllers without hardware blinking capability could use it (and not only led_gpio). If the feature is enabled, then led_blink() could run the hardware method 'api->blink()' if available and fallback on the software implementation else.

simonguinot reviewed (changes requested):
Hi @KAGA164 
Interesting work. Is this PR still alive ?

KAGA164 commented:
Hi @simonguinot, Thanks for comments. I hope that It will be still possible to merge this PR. My idea was to generate static initializer for every LEDs from DTS and initialize a driver with this data to avoid create driver instance for any sigle LED to achieve that I requested some changes in DTS #21393 but It did not be merged. Since we have a new DTS API I still do not see a possibility to get all LED data or iterate over all LEDs but I migth be wrong. Maybe @pabigot, @ulfalizer or @carlescufi could comment if something like this driver needed is possible now. As you write in comments It would be nice to have possibility to additionally group LEDs in group or controllers. This driver handle GPIO LEDs and leds is chosen by the additional parameters in DTS "id" so can be connected to default gpio port or pin and LEDs count is limited to `u32_t` max value.

carlescufi commented:
@KAGA164 could you please rebase?
@mbolivar-nordic could you review this one?

mbolivar-nordic reviewed (commented):
The main thing I think needs doing is a rebase and conversion onto the new devicetree API. This is currently based on an abandoned devicetree change which added static initializers for entire nodes.

I think making this multi-instance would be nice as a future enhancement, but if we could start by supporting the single `/leds` node that a lot of boards define, it would be a good enough beginning. It should be straightforward to convert to a DT_INST_FOREACH with DT_DRV_COMPAT set to gpio_leds after that.

I agree with comments I've seen elsewhere (I think from @simonguinot ?) that centralizing software blinking in a core leds.c file instead of doing it per-driver would be nicer.

I'm not sure why the ID is needed, but I didn't read the entire description. Assuming that we go with something like https://github.com/zephyrproject-rtos/zephyr/pull/26101, then perhaps the children of the gpio-leds node could simply be placed into a static array using DT_FOREACH_CHILD in the new DT API. The index of each child in the array could be the led number in the 26101 way of doing things. But that would need some discussion, I think @arvid-r has another approach where all the GPIO LEDs owned by a particular GPIO controller are grouped together?

In all I think there's good ideas here, just needs driving to consensus. I'm happy to help support on the devicetree API side of things.

KAGA164 commented:
Thanks @mbolivar.  The `DT_FOREACH_CHILD `is what I need here. Probably all leds(child) number will be needed also to reserve the appropriate amount of static memory for leds data. The #26101 provides the led_controller which is a good step forward and contains all needed data for my conceptions. The LED ID is exactly the same things like index in #26101. I use it to distinguish diodes because under `/leds` node LEDs can have the same GPIO Pin.

Summarizing we have two possible solution:
- At the beginning we can support only leds from `/leds`. There is `DT_FOREACH_CHILD` API, which probably can be used to retrieve all definied LEDs data. We also need information about leds count, it can be added to DTS API as for example `DT_CHILD_NUM` or maybe this is available already. Could you confirm @mbolivar ?

- Second options can be done after #26101 goes in or in feature after option one. Then we can create this driver instance per each LEDs controller where LEDs are grouped under each controller. 

After deciding what to do next I can rebase PR and resume work on it taking into account the comments .

mbolivar-nordic commented:
> Thanks @mbolivar.

Sure thing!

By the way, please use @mbolivar-nordic for work tasks, if you don't mind. I no longer use @mbolivar for work and I may not check my notifications there very quickly. Thanks :smile: 

>     * At the beginning we can support only leds from `/leds`. There is `DT_FOREACH_CHILD` API, which probably can be used to retrieve all definied LEDs data. We also need information about leds count, it can be added to DTS API as for example `DT_CHILD_NUM` or maybe this is available already. Could you confirm @mbolivar ?

I think that would be a solid start. We can add any devicetree APIs necessary, including DT_NUM_CHILDREN or so. Just let me know. If you are capturing all the children in an array, you can also use an idiom which combines ARRAY_SIZE with DT_FOREACH_CHILD to count the number of LED children you care about as is done in #26101.

I think @simonguinot is saying in https://github.com/zephyrproject-rtos/zephyr/pull/26101#discussion_r443530580 that this gpio-leds driver would probably not use the led-controller.yaml file, but maybe they can confirm.

KAGA164 commented:
@mbolivar-nordic I successfully rebased this PR and I was able to retrive all needed LED data using the new DT API. This driver still uses the ID property to identify the LED it seems for me it is ok since we do not create a driver instance for each GPIO port or virtual LED controller. According to @simonguinot comment about doing software blink functionality to common led.c I think that this functionality is dedicated to driving LEDs by GPIO and other hardware LEDs controllers probably needs to implement their own software blink. It seems, like creating the common software blink migth be really hard to implementation and may require many changes to the API.

carlescufi commented:
@simonguinot are you OK with merging this now?

pabigot commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,279 @@
+
+	data->state = state;
+
+	if (data->cfg->gpio.polarity & GPIO_OUTPUT_ACTIVE) {
```
Remove all this.  It should be handled by `gpio_pin_configure()`.

KAGA164 replied:
Removed

pabigot commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,279 @@
+#define LED_SW_BLINK BIT(0)
+#define LED_BLINK_DISABLE BIT(1)
+
+struct gpio_cfg {
```
The `pin` field here should be type `gpio_pin_t`.  The `polarity` flag here should be named `flags` (or `dt_flags`) and be of type `gpio_dt_flags_t`.

I don't see why this is a separate structure rather than part of `led_dts_cfg`.

KAGA164 replied:
I changed `pin`->`gpio_pin_t` and `polarity` -> `gpio_dt_flags_t`. I used a separate structure for readability

pabigot commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,279 @@
+	}
+
+	err = gpio_pin_set(data->gpio_dev, data->cfg->gpio.pin,
+			   pin_value);
```
for `pin_value` use `state`.

pabigot commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,279 @@
+
+		/* Set all leds pin as output */
+		err = gpio_pin_configure(data[i].gpio_dev, cfg->led[i].gpio.pin,
+					 GPIO_OUTPUT);
```
This should be
```
		err = gpio_pin_configure(data[i].gpio_dev, cfg->led[i].gpio.pin,
￼					 GPIO_OUTPUT_INACTIVE | cfg->led[i].gpio.dt_flags);
```

KAGA164 replied:
Done

pabigot commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,279 @@
+			return err;
+		}
+
+		if (cfg->led[i].gpio.polarity & GPIO_OUTPUT_ACTIVE) {
```
and this goes away.

pabigot commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,279 @@
+			pin_value = 1;
+		}
+
+		err = gpio_pin_set(data[i].gpio_dev, cfg->led[i].gpio.pin,
```
so does this.

pabigot commented on dts/bindings/gpio/gpio-leds.yaml:
```diff
@@ -15,3 +22,7 @@ child-binding:
           required: true
           type: string
           description: Human readable string describing the device (used as device_get_binding() argument)
+       id:
```
This is an API change, incompatible with Linux where `gpio-leds` is defined.

pabigot replied:
Also the example devicetree modifications use identifiers that start with 1, which is unusual.  Is `id` intended to be an opaque label, or an ordinal identifying the LED based on position in some ordering?

Finally the Zephyr `led` API requires this to be interpreted as an unsigned value, which should be noted (if this property is acceptable in this compatible).

KAGA164 replied:
Why we need to be compatible with the Linux ? I do no have to start from 1, I did this by chance. ID is used to identify the LED by API, because the GPIO port can be different for every LEDs in the `leds` node so the LEDs can have the same GPIO Pin so we cannot drive LEDs using it. That is why I provide a ID property which only should be unique for each LED. ID is used in the LED API to control the leds for_example `led_on(led, id)`. Do you have any idea how to get ride with this ID property without changing whole led API and maintaining compatibility with Linux ?

pabigot replied:
I don't believe there's any guarantee that the order of children is preserved across builds, but that would probably be least intrusive way to go.

simonguinot replied:
Arf. I am always forgetting this one.

I think it is possible to make this index/ID "optional" and then to keep compatibility with the existing DTS coming from Linux. If the property is available in the LED subnode, then the led-gpio driver could use it. This would allow to ensure the LED order. But if the property is not available then the index of the LED in the array could be used. [26101](https://github.com/zephyrproject-rtos/zephyr/pull/26101) introduces a led_get_info syscall. This syscall allows to retrieve information (label, index, ...) for a given LED number. If this callback is provided by the led_gpio driver then a user would still be able to discover which LED label is mapped to which LED number (if the ID to enforce it is missing).

And to finish, if an ID/index is required, then obviously I would like to see the index led-controller property from [26101](https://github.com/zephyrproject-rtos/zephyr/pull/26101) used :)

KAGA164 replied:
At this moment this ID property is optional but I assumed that if you not defining it in dts, you should not use this driver. Since we do not have this `led_get_info` functionality removing id and basing on index which depends on the leds order in board dts is a little risky like @pabigot said there is not guarantee that somebody do not change the leds order in dts. This risk is probably very small.

simonguinot replied:
Well, the "id" property is required to use the led-gpio driver. Without this property, the driver won't work. So it should not be marked as optional in the binding because it is not. But if you allow the driver to work even if the "id" is not provided (by using indexes) then the "id" property becomes truly optional. And the driver becomes compatible with the existing board DTS from Linux as well.

Anyway, I think it would be nice to have this PR based on [26101](https://github.com/zephyrproject-rtos/zephyr/pull/26101). This would provide the led_get_info syscall and the index property.

mbolivar-nordic replied:
> Why we need to be compatible with the Linux ?

This is a documented design goal:

https://docs.zephyrproject.org/latest/guides/dts/design.html#source-compatibility-with-other-operating-systems

> I don't believe there's any guarantee that the order of children is preserved across builds, but that would probably be least intrusive way to go.

What if we just added a new devicetree API macro named `DT_CHILD_INDEX(node_id)` which let you get the index of `node_id` in its sibling list?

So for example, if this is the dts:

```
        leds {
                compatible = "gpio-leds";
                red_led: led_0 {
                        gpios = < &gpiob 0x16 0x1 >;
                        label = "User LD1";
                };
                green_led: led_1 {
                        gpios = < &gpioe 0x1a 0x1 >;
                        label = "User LD2";
                };
                blue_led: led_2 {
                        gpios = < &gpiob 0x15 0x1 >;
                        label = "User LD3";
                };
        };
```

then:

- `DT_CHILD_INDEX(DT_NODELABEL(red_led))` is 0
- `DT_CHILD_INDEX(DT_NODELABEL(green_led))` is 1
- `DT_CHILD_INDEX(DT_NODELABEL(blue_led))` is 2

If I understand correctly, that at least is the necessary mechanism.

Or do people just want to wait for and rebase on top of #26101?

mbolivar-nordic replied:
> . The index in the `leds[]` array provides the id that identifies that LED.

I see how that takes you from a parent node an array initializer.

What I don't follow is how to start with a devicetree node referring to a child (i.e. referring to an LED), and get an index.

pabigot replied:
I guess it depends on how this is supposed to be used.  But you're right, if you want to be able to identify a specific node by its devicetree name, then `DT_CHILD_INDEX()` is the way to do it.

pabigot commented on dts/bindings/gpio/gpio-leds.yaml:
```diff
@@ -5,6 +5,13 @@ description: GPIO LEDs parent node
 
 compatible: "gpio-leds"
 
+properties:
```
This does not belong here either.  If something like this were needed, it'd be called `label`.

KAGA164 replied:
It can be renamed to `label` or even moved to Kconfig. I would to define here the driver name which can be used for bindings in `device_get_binding`

pabigot replied:
The standard way to do that is:
```
include: base.yaml

properties:
    label:
      required: true
```

mbolivar-nordic replied:
Agree with @pabigot, this needs to be done with the 'label' property. We have documentation for this here that users rely on:

https://docs.zephyrproject.org/latest/guides/dts/howtos.html#get-a-struct-device-from-a-devicetree-node

pabigot reviewed (changes requested):
This goes too far changing what `gpio-leds` means as a compatible.

pabigot commented:
@galak Please review this from the perspective of `gpio-leds` as a compatible and how a driver should be associated with it.

simonguinot reviewed (commented):
Hi @KAGA164,

Thanks for this update. It is good to see this PR is moving on.

My major comment is about the software blinking support. I still think that implementing software blinking in LED drivers is not a good idea. IMHO a driver should only provide the blink callback if some hardware capabilities are available. Else a software blinking function could be used as a fallback. This way the same software blinking code would be shared between all the drivers. It is how it is done in Linux and I think it is the right way to go.

Please see the following comments as well.

simonguinot commented:
> @simonguinot are you OK with merging this now?
> @mbolivar-nordic please take another look

Hi @carlescufi. It is not my place to say but thanks for asking. I think there are good things in this PR but I would not merge it right now. I would wait for reaching an agreement on the "id" property thing. I believe it is still possible. I also have a disagreement about software blinking. I don't think it should be implemented in the led-gpio driver (or any other LED driver). In addition, I sent a couple of comments to @KAGA164.

KAGA164 commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,168 @@
+	.led = led_dts
+};
+
+DEVICE_AND_API_INIT(gpio_led, DT_PROP(DT_DRV_INST(0), driver),
```
I don't mind many instances of this driver. I just wonder if it's really needed. I can add support for it if you want to. Is it possible to have many `leds` node instance ?

simonguinot replied:
I think it is not super useful but I don't see any reason for not supporting it.

As an example, let's imagine a board with two distinct arrays of LEDs. If each array can have its own led-gpio node, then the LED indexes (or IDs) will reflect their hardware position in the array. It could be more meaningful for a user than having a big single node with the LED indexes not related with the hardware.

mbolivar-nordic replied:
In my opinion we can just support one instance to start with and make the driver multi-instance later on if we need to, since we don't have a use case for more right now.

Nothing will break and it will be totally backwards compatible as long as we access the devices using label properties as discussed elsewhere in these comments.

KAGA164 commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,168 @@
+
+static const struct led_driver_api led_api = {
+	.on = gpio_led_on,
+	.off = gpio_led_off,
```
I can define it but I would not remove this on/off callback I rather prefer to use it instead of `set_brightness` knowing that I use the GPIO LED driver where LED can be only on or off. It is less confusing for me

simonguinot replied:
Of course. I am just saying you could add the `set_brightness` callback.

mbolivar-nordic replied:
Maybe just add a callback that returns -ENOTSUP? The led.h API, at least as it is currently defined, requires set_brightness to be defined. So this is technically an API violation, even though I agree with @KAGA164 that it doesn't make sense for an LED connected to a GPIO to do anything with set_brightness.

KAGA164 commented on drivers/led/led_gpio.c:
```diff
@@ -92,6 +101,18 @@ static int gpio_led_set_no_delay(struct led_gpio_data *data, bool state)
+		data->next_state = state;
+
+		atomic_set_bit(&data->flags, LED_BLINK_DISABLE);
+		k_timer_start(&data->timer, K_NO_WAIT, K_NO_WAIT);
```
This is here to avoid some race conditions when during the software blink is used, the led migth be set to different state from a higher priority interrupt or thread than a timer handler also the set function can be called when timer is pending. So setting the LEDs from one context and using the atomic flags eliminates this problem.

KAGA164 commented on drivers/led/led_gpio.c:
```diff
@@ -36,6 +38,13 @@ struct led_dts_cfg {
 	struct device *gpio_dev;
 	const struct led_dts_cfg *cfg;
+#if defined(CONFIG_LED_SOFTWARE_BLINK)
+	struct k_timer timer;
```
It can be a problem. Maybe it was not possible to sleep in the `gpio_pin_set` or I did not realize that it is possible. Probably in that case we have to use a timer instance and workqueue instance for software blink but if this function can sleep then we cannot use also the system workqueue because it is also not good idea to sleep there so we have to define own workqueue for software blink feature which means creating the new thread for me it would be some overkill in used resource.

I will check, how Linux handle LEDs. I would like to have a possibility to call LED API from irq context.

BTW There is no any word in `gpio_pin_set` that this function should not be used from IRQ context or that it can sleep.
Do we have any cases where I2C GPIO expander is under it ?

KAGA164 replied:
Ok, we have SPI/I2C GPIO expander driver which works synchronously

KAGA164 replied:
I checked the Linux. Just as I thought it used combination of software timer and workqueue to handle software blink.

KAGA164 replied:
So since I know `gpio_pin_set` can be blocking we probably need to do some simillar logic here or like you said we can move this to something like generic ` led.c`

simonguinot replied:
I have been thinking more to this sleeping context issue.

Here is how it could be handled:

We could introduce some _nosleep variants for LED API functions and also a "can_sleep" flag to advertise if a LED device/driver can sleep or not. If a user calls a classic variant (e.g. led_on), then it means that the driver is allowed to sleep. The driver callback can be run directly. But if a user calls the _nosleep variant (e.g. led_on_nosleep), then the LED system has to check if the driver may sleep or not. If not then again the driver callback can be run directly. Else a work queue can be used to defer the work into a thread context. A timer handler (as for software blinking) should use the _nosleep variants.

So I think it is correct to use timers for LED software blinking. We don't want the work queue overhead if we can avoid it. The "sleeping" context issue is already here and the fix should not prevent from configuring LEDs from timers.

As a side note, AFAIK the "can_sleep" information is not exported/supported by the GPIO subsystem yet. At some point it may be needed by the led-gpio driver.

jfischer-phytec-iot replied:
Maybe "Asynchronous sequence manager #23075" can be used for LED software blinking. cc @nordic-krch

nordic-krch replied:
that wouldn't not fit here but if gpio expander would use i2c_async (#23371) then it could be called from interrupt context.

pabigot replied:
> As a side note, AFAIK the "can_sleep" information is not exported/supported by the GPIO subsystem yet. At some point it may be needed by the led-gpio driver.

The GPIO API is documented to return `-EWOULDBLOCK` when invoked from a context where the operation can't be completed synchronously.  This is implemented in the SX1509B driver, and I believe in subsequent IOX devices.

I don't see how `i2c_sync` would help with invoking synchronous GPIO calls from an interrupt.  Though if the `-EWOULDBLOCK` return value were detected the operation could be re-initiated with an async GPIO API, but that API would have to be designed.

Seems a lot more complex than a work queue, though.

nordic-krch replied:
> I don't see how i2c_sync would help with invoking synchronous GPIO calls from an interrupt

They would not be synchronous. Using work queue adds much more jitter and latency compared to timer. From on chip gpios this can be seen as unnecessary limitation.

KAGA164 replied:
If the GPIO API can return the `-EWOULDBLOCK` when invoked from a context where the operation can't be completed synchronously. I would suppose that timer handler is such context then we can call it again by triggering a workqueue from timer context and there we can call the GPIO API. For pure GPIO driving the GPIO API will be called directly in timer handler because it does not return `-EWOULDBLOCK` so I think that it will not be a limitation there. From the other side it would not be a good idea to use the system workqueue because blocking there have an impact on other subsystem for example Bluetooth. We can define here a workqueue instance which will be used only by this feature but it means creating a new thread which will significantly increase the volatile memory consumption for such a simple functionality like LED blinking.

pabigot replied:
Bluetooth should not be affected by items in the system work queue unless those items misbehave.  I don't think this driver should define a workqueue, but agree we need better control of which work queue is to be used by drivers that need them.

simonguinot replied:
Again, I think that software LED blinking should be a fallback for all the LED devices/drivers without hardware blinking capabilities, and not only GPIO LEDs.

Moreover context should be handled from the led_set_{on,off,brightness} functions. Software LED blinking is only a user among others. To make a decision between calling the LED driver's callback directly or defer it to a workqueue, this functions need to be aware of the context (interrupt, thread) and also if the driver may sleep. I don't think that calling the driver's callback and handling a -EWOULDBLOCK error is correct. GPIO is not the only backend we have to deal with. It would be better to have a "can_sleep" flag advertised by the LED drivers. But to be efficient this needs some support from all the possible backends (including GPIO)...

pabigot replied:
>  I don't think that calling the driver's callback and handling a -EWOULDBLOCK error is correct. GPIO is not the only backend we have to deal with. It would be better to have a "can_sleep" flag advertised by the LED drivers. 

GPIO is not the only API where an operation may not be [isr-ok](https://docs.zephyrproject.org/latest/reference/terminology.html#api-term-isr-ok).  The standard way of nicely supporting code that tries something that can't be done in the calling (interrupt) context is to return `-EWOULDBLOCK`.  This fast-exit should have minimal performance impact, essentially no different to retrieving and checking a flag.

I understand that this seems a little odd, but it is the most flexible/least impact solution we have available.  Since this LED driver can't tell whether a particular GPIO has to support can change pin state from interrupt context, so there is no way it could advertise a "can sleep" flag.

mbolivar-nordic commented on drivers/led/led_gpio.c:
```diff
@@ -0,0 +1,168 @@
+	{                                      \
+		DT_GPIO_LABEL(node_id, gpios), \
+		DT_GPIO_PIN(node_id, gpios),   \
+		DT_GPIO_FLAGS(node_id, gpios)  \
```
Tricky: not every GPIO specifier has a flags field in devicetree, although almost all of them do. It's still a portability bug for a generic driver like this one to assume there is a flags field.

For a way to handle this, see the FLAGS_OR_ZERO macro from samples/subsys/usb/hid-cdc/src/main.c. I will try to implement https://github.com/zephyrproject-rtos/zephyr/issues/25827 after reviewing this to pull out the generic code if that helps.

mbolivar-nordic replied:
> Tricky: not every GPIO specifier has a flags field in devicetree

https://github.com/zephyrproject-rtos/zephyr/pull/27076 resolves this comment; feel free to ignore this comment as I expect this will be merged

mbolivar-nordic commented on include/drivers/led.h:
```diff
@@ -73,7 +73,7 @@ __subsystem struct led_driver_api {
  *
  * @param dev LED device
- * @param led LED channel/pin
+ * @param led LED channel/pin/id
```
This part I think conflicts with #26101 here:

https://github.com/zephyrproject-rtos/zephyr/pull/26101/commits/db8213c76ed2414be464a902c0af4495206840f3

(The shortlog is ` led: move API functions to LED objects` in case that commit SHA goes stale due to rebasing.)

I honestly like the #26101 phrasing a bit better. Could we adopt that here?

mbolivar-nordic commented:
@KAGA164 thanks very much for your update. I think it's looking much better. I have some comments and ideas I've left above. Sorry for the long review time; I was working on downstream tasks.

simonguinot commented:
Hi @KAGA164. Are you still working on this PR ?

Mani-Sadhasivam commented:
@KAGA164 Is this PR still active?

KAGA164 commented:
Hi @Mani-Sadhasivam @simonguinot Sorry for the late answer. Unfortunately, I haven't had much time to work on this PR lately. I have read all of the comments and have some conclusions:

1. The software blink functionality can be moved to separated file and can be used as fallback for all LEDs drivers if those drivers don't set a led_blink(if set it means hardware blink is supported) API. It would need to  sw timer and workqueue instance be added to every led data structure if software blink is enabled.
2. Since the GPIO API or other led API can block. We would need to have a software timer and workqueue instance per each led. Where a software timer is used to measure a delay_on/delay_off time. Next the led_on/led_off API is called from a timer handler(irq context) when it returns ` -EWOULDBLOCK` then the workqueue will be used to call led_on/led_off from no interrupt context.
3. This driver was supposed to support the boards led devicetree node which based on the `gpio-leds.yaml` bindings since I cannot add there a id/index property like I did in this PR. How can we use a LED API to to select the correct LED ? Example of current LED node:
```
        leds {
                compatible = "gpio-leds";
                red_led: led_0 {
                        gpios = < &gpiob 0x16 0x1 >;
                        label = "User LD1";
                };
                green_led: led_1 {
                        gpios = < &gpioe 0x1a 0x1 >;
                        label = "User LD2";
                };
                blue_led: led_2 {
                        gpios = < &gpiob 0x15 0x1 >;
                        label = "User LD3";
                };
        };
```
Can we add `DT_CHILD_INDEX` macro do the devicetree as @mbolivar-nordic suggest ?
And then use:

```
    DT_CHILD_INDEX(DT_NODELABEL(red_led)) is 0
    DT_CHILD_INDEX(DT_NODELABEL(green_led)) is 1
    DT_CHILD_INDEX(DT_NODELABEL(blue_led)) is 2
```

There is also new bindings for led controller(led-controller.yaml) which would be suitable for this driver and then we can also add support for multi-instance but how can we control a LEDs from the boards devicetree then ? Should we add a led_controller to every board which has a leds node instead of it ?

What do you think about this @Mani-Sadhasivam @simonguinot @mbolivar-nordic ?

mbolivar-nordic commented:
@KAGA164 I am happy to add DT_CHILD_INDEX if that's useful, yes.
